### ISSUE TITLE
Add support for Cypress 10

### ISSUE THREAD

**Issue by tgamauf** [tgamauf/cypress-parallel#2]
Cypress 10 changes it's config from JSON to `.config.js/.ts/.mjs/.cjs` and removes JSON support. This complicates things significantly, especially for Typescript configs as these cannot be read easily in the compiled Javascript that is used in the action. I need to find a way to accomplish this in order to make the action useful for future Cypress versions.

Cypress 10 migration guide: https://docs.cypress.io/guides/references/migration-guide#Configuration-File-Changes
Cypress 10 config: https://docs.cypress.io/guides/references/configuration#Configuration-File

**Comment by mehreenhamraz** [comment]
Hi Any update on this?

**Comment by tgamauf** [comment]
Sorry, I currently have no time to implement this. But I'm happy to accept pull requests!

**Comment by Manuel-Suarez-Abascal** [comment]
@tgamauf first of all, thanks a lot for this plugin. It has allowed me to run our Cypress e2e test suites in parallel using scheduled run on CI using GitHub Actions. Unfortunately, we're stuck when updating to Cypress 10 due to this dependency. Are there any updates on this issue?

I would be willing to make a Pull Request, but I will need some assistance on how to contribute to it. Can you offer me some pointers on which files are responsible for what? What are the challenges? How to test if my possible fix is working?...etc...

**Comment by tgamauf** [comment]
Sure, I would be happy for a solution!

There really is only a single file that is relevant for this: https://github.com/tgamauf/cypress-parallel/blob/283d8935c81e7eb304238e080019470cca7a9c4a/src/parsing.ts. Two parts have to be adapted:
1. the function `findWorkingDirectory` looks for the Cypress config file. At the moment this will always look for the `cypress.json` file, but now has to look for all possible versions: `cypress.config.js`, `cypress.config.ts`, `cypress.config.mjs`, `cypress.config.cjs`. Another issue here is if there are multiple config files found, but this isn't valid according to the migration guide anyway
2. the function `loadCypressConfig` is responsible for parsing the config file and providing the configuration values that are relevant for us. Up to now this was `componentFolder`, `ignoreTestFiles`, `integrationFolder`, `testFiles`. However, this changed to `specPattern` and `excludeSpecPattern`. Also, there might be more changes that I haven't found yet - need to check the migration guide for that as well.

Testing is pretty simple, you can find everything test related in `__tests__` - simply replace the current fixtures with fixtures for all of the different config file types and change and extend the tests in https://github.com/tgamauf/cypress-parallel/blob/283d8935c81e7eb304238e080019470cca7a9c4a/__tests__/parsing.test.ts.

The main problem here is this: up to version 10, Cypress configuration was done by the `cypress.json` file in the Cypress root. As it was JSON it was trivial to read with Javascript. In version 10 they changed that to a `cypress.config.js` (and a few other file endings), which doesn't contain static key-value pairs, but Javascript/Typescript code. Something like that:
```javascript
const { defineConfig } = require('cypress')

module.exports = defineConfig({
  e2e: {
    baseUrl: 'http://localhost:1234'
  }
})
```
or something very similar in Typescript.

I don't see any easy way to reliably parse this myself. As this can contain functional JS/TS code I basically would have to implement an interpreter that reads the file, "understands" it, and then provide the few config values that are actually relevant for cypress-parallel. My preferred solution would be to use the original Cypress code - there has to be some function in the Cypress code that actually reads this config and it should be possible to reuse this. Most likely it isn't public, which means that we would have to copy the code (which is rarely a good approach, but better than nothing). As I don't know the Cypress internals well, I couldn't find the parts responsible for parsing the config when I looked.

I also know that a number of widely used JS projects like Webpack use this or a very similar JS config scheme, so it is possible that a general "Javascript config parsing library" exists. I couldn't find anything when I looked, though.

**Comment by tgamauf** [comment]
Btw. backward compatibility doesn't matter - this version of the action will remain the version up to Cypress 10.

**Comment by JulianRoesner** [comment]
I just ran into the problem when trying to use this action in one project where Cypress 10 was already in use. I was wondering whether it would be a temporary workaround to add a parameter `test-location` to the action. This could then be used to directly specify the folder and skip the parsing of the configuration file. Or are there other parts taken from the configuration?

**Comment by tgamauf** [comment]
As mentioned above, the following Cypress config options are currently used: `componentFolder`, `ignoreTestFiles`, `integrationFolder`, and `testFiles`. In Cypress 10 the corresponding options are `specPattern` and `excludeSpecPattern`. This allows filtering of test specs. So sadly it‘s not as simple as that …

**Comment by tgamauf** [comment]
Alright, guys, sorry that this took so long. But Cypress 10+ should work now.

Thanks to @kgw4it for pointing me in the right direction and starting the implementation!
